c2rust-refactor/scripts/upgrade_ptr_to_ref.lua

```diff
@@ -0,0 +1,708 @@
+        -- TODO: Should probably expand to work on more complex exprs
+        elseif rhs_kind == "Path" then
+            local hirid = self.tctx:resolve_path_hirid(rhs)
+            local var = self:get_var(hirid)
+
+            if var and not self.node_id_cfgs[var.id]:is_opt_any() then
+                local lhs_ty = self.tctx:get_expr_ty(lhs)
+
+                -- If lhs was a ptr, and rhs isn't wrapped in some, wrap it
+                -- TODO: Validate rhs needs to be wrapped
+                if lhs_ty:get_kind() == "Ptr" then
+                    local some_path_expr = self.tctx:ident_path_expr("Some")
+
+                    rhs:to_call{some_path_expr, rhs}
+                    expr:set_exprs{lhs, rhs}
+                end
+            end
+        end
+    -- free(foo.bar as *mut libc::c_void) -> foo.bar.take()
+    elseif expr_kind == "Call" then
+        local call_exprs = expr:get_exprs()
+        local path_expr = call_exprs[1]
+        local param_expr = call_exprs[2]
+        local path = path_expr:get_path()
+        local segments = path:get_segments()
+
+        -- In case free is called from another module check the last segment
+        if segments[#segments] == "free" and param_expr:get_kind() == "Cast" then
+            -- REVIEW: What if there's a multi-layered cast?
+            local uncasted_expr = param_expr:get_exprs()[1]
+            local conversion_cfg = self:get_expr_cfg(uncasted_expr)
+
+            if conversion_cfg and conversion_cfg:is_opt_any() then
+                expr:to_method_call("take", {uncasted_expr})
+            end
+        -- ip as *mut c_void -> ip.as_mut_ptr() as *mut c_void
+        -- Though this should be expanded to support other exprs like
+        -- fields
+        elseif segments[#segments] == "memset" then
+            param_expr:map_first_path(function(path_expr)
+                local cfg = self:get_expr_cfg(path_expr)
+
+                if cfg and cfg:is_box_any() then
+                    path_expr:to_method_call("as_mut_ptr", {path_expr})
+                end
+
+                return path_expr
+            end)
+
+            call_exprs[2] = param_expr
+
+            expr:set_exprs(call_exprs)
+        end
+    end
+end
+
+function Visitor:get_expr_cfg(expr)
+    local hirid = self.tctx:resolve_path_hirid(expr)
+    local node_id = nil
+    local var = self:get_var(hirid)
+
+    -- If we're looking at a local or param, lookup from the variable map
+    if var then
+        node_id = var.id
+    -- Otherwise check the field map
+    elseif expr:get_kind() == "Field" then
+        hirid = self.tctx:get_field_expr_hirid(expr)
+        local field = self:get_field(hirid)
+
+        if field then
+            node_id = field.id
+        end
+    end
+
+    return self.node_id_cfgs[node_id]
+end
+
+-- HrIds may be reused in different functions, so we should clear them out
+-- so we don't accidentally access old info
+function Visitor:visit_fn_decl(fn_decl)
+    self.vars = {}
```

Comment:
Technically functions can be nested in rust, so this would clear in the middle of processing a function if so.

Reply:
That's a good point. I don't think we'll see much of that in translated code. But certainly should be considered.

Wouldn't those be two separate visits? IE visit outer then visit inner? Or would the inner item get visited mid outer visit?

Reply:
Inner would get visited inside the outer.

Reply:
Interesting. I wonder if I'll need like a stack of variables to push and pop or something

Reply:
That's what I've done for nested modules. It would be nice to assert or something on this case, but I don't think it's critical. Needs a least a fat warning comment.